software/bsg_manycore_lib/bsg_mcs_mutex.S

```diff
@@ -0,0 +1,81 @@
+// MCS mutex
+// Author: Max
+//
+// This is an implementation of the MCS mutex described by Mellor-Crummey and Scott in their 1991 paper
+// “Algorithms for Scalable Synchronization on Shared-Memory Multiprocessors”
+//
+// This is a spinlock mutex, but unlike a simple spinlock in which all threads update and spin on
+// a single memory location, the MCS lock builds a linked-list of memory locations local to each core.
+//
+// Cores atomically append their local memory region to the global list using an unconditional
+// amoswap operation. They then spin on their local memories for a predecessor in the queue
+// to notify them that they now hold the lock.
+//
+// Once a core has completed its critical region, it checks for a successor and updates releases the lock to them.
+//
+// The advantages of this mutex over a simple spin lock on the manycore are two fold:
+//
+// (1) It greatly reduces the number of memory reqeusts on the network and it mitigates the extent to which
+// a single memory bank becomes a hot-spot. The number of requests issued to a memory bank containing the
+// lock object is linear with the number of times an acquire operation is execution.
+//
+// (2) The lock approximates a FIFO-ish structure, which improves fairness. A simple spinlock on the manycore
+// will favor threads topologically closer to the memory bank in which the lock resides and can lead to
+// starvation of the other cores.
+//
+// This lock is by no means perfect. For locks with low contention, a simple spinlock may result in better performance.
+
+        .text
+        .globl bsg_mcs_mutex_acquire
+bsg_mcs_mutex_acquire:
+        // a0 = mtx         : bsg_mcs_mutex_t*
+        // a1 = lcl         : bsg_mcs_mutex_node_t*
+        // a2 = lcl_as_glbl : bsg_mcs_mutex_node_t*
+        sw      x0, 0(a1)                      // lcl->next = 0
+        sw      x0, 4(a1)                      // lcl->unlocked = 0
+        amoswap.w.aq t0, a2, 0(a0)             // predecessor = swap (&mtx, lcl_as_glbl)
+        addi    t1, a1, 4                      // t1 = &lcl->unlocked
```

Comment:
Is it possible to swap the location of next and unlocked in bsg_mcs_mutex_node_t, so you can get rid of this addi?

Reply:
I think this is the dual of the comment I made below - I asked about compiler reordering. If the struct is reordered, then there needs to be an addi in bsg_mcs_mutex_release_no_successor to access the next pointer. Other places may exist, but that's where I put my comment.

Reply:
Well currently I don't do the load on broken reservation when spinning on the next pointer in the no successor case, so I can just use an offset load to access the next pointer. I commit the load-on-broken because we can expect that pointer to be updated imminently by the successor that's just added themselves to the list.